File: .NET/Microsoft.Recognizers.Text.DataTypes.DataDrivenTests/TestTimex.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Microsoft.Recognizers.Text.DataTypes.TimexExpression.Tests
{
    [TestClass]
    public class TestTimex
    {
        [TestMethod]
        public void DataTypes_Timex_FromDate()
        {
            Assert.AreEqual("2017-12-05", TimexProperty.FromDate(new System.DateTime(2017, 12, 5)).TimexValue);
        }

        [TestMethod]
        public void DataTypes_Timex_FromDateTime()
        {
            Assert.AreEqual("2017-12-05T23:57:35",
                TimexProperty.FromDateTime(new System.DateTime(2017, 12, 5, 23, 57, 35)).TimexValue);
        }

        [TestMethod]
        [DataRow("2017-09-27")]
        [DataRow("XXXX-WXX-3")]
        [DataRow("XXXX-12-05")]
        public void DataTypes_Timex_Roundtrip_Date(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("T17:30:45")]
        [DataRow("T05:06:07")]
        [DataRow("T17:30")]
        [DataRow("T23")]
        public void DataTypes_Timex_Roundtrip_Time(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("P50Y")]
        [DataRow("P6M")]
        [DataRow("P3W")]
        [DataRow("P5D")]
        [DataRow("PT16H")]
        [DataRow("PT32M")]
        [DataRow("PT20S")]
        public void DataTypes_Timex_Roundtrip_Duration(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("PRESENT_REF")]
        public void DataTypes_Timex_Roundtrip_Now(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("XXXX-WXX-3T04")]
        [DataRow("2017-09-27T11:41:30")]
        public void DataTypes_Timex_Roundtrip_DateTime(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("2017")]
        [DataRow("SU")]
        [DataRow("2017-WI")]
        [DataRow("2017-09")]
        [DataRow("2017-W37")]
        [DataRow("2017-W37-WE")]
        [DataRow("XXXX-05")]
        public void DataTypes_Timex_Roundtrip_DateRange(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("(XXXX-WXX-3,XXXX-WXX-6,P3D)")]
        [DataRow("(XXXX-01-01,XXXX-08-05,P216D)")]
        [DataRow("(2017-01-01,2017-08-05,P216D)")]
        [DataRow("(2016-01-01,2016-08-05,P217D)")]
        public void DataTypes_Timex_Roundtrip_DateRange_Start_End_Duration(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("TEV")]
        public void DataTypes_Timex_Roundtrip_TimeRange(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("(T16,T19,PT3H)")]
        public void DataTypes_Timex_Roundtrip_TimeRange_Start_End_Duration(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("2017-09-27TEV")]
        public void DataTypes_Timex_Roundtrip_DateTimeRange(string timex)
        {
            Roundtrip(timex);
        }

        [TestMethod]
        [DataRow("(2017-09-08T21:19:29,2017-09-08T21:24:29,PT5M)")]
        [DataRow("(XXXX-WXX-3T16,XXXX-WXX-6T15,PT71H)")]
        public void DataTypes_Timex_Roundtrip_DateTimeRange_Start_End_Duration(string timex)
        {
            Roundtrip(timex);
        }

        private static void Roundtrip(string timex)
        {
            Assert.AreEqual(timex, new TimexProperty(timex).TimexValue);
        }
    }
}

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/test/java/com/microsoft/recognizers/text/datatypes/timex/expression/TestTimex.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.text.datatypes.timex.expression;

import com.microsoft.recognizers.datatypes.timex.expression.Time;
import com.microsoft.recognizers.datatypes.timex.expression.TimexProperty;

import java.time.LocalDateTime;

import org.junit.Assert;
import org.junit.Test;

public class TestTimex {
    @Test
    public void dataTypesTimexFromDate() {
        LocalDateTime date = LocalDateTime.of(2017, 12, 5, 0, 0);
        Assert.assertEquals("2017-12-05", TimexProperty.fromDate(date).getTimexValue());
    }

    @Test
    public void dataTypesTimexFromDateTime() {
        LocalDateTime date = LocalDateTime.of(2017, 12, 5, 23, 57, 35);
        Assert.assertEquals("2017-12-05T23:57:35", TimexProperty.fromDateTime(date).getTimexValue());
    }

    @Test
    public void dataTypesTimexRoundtripDate() {
        TestTimex.roundtrip("2017-09-27");
        TestTimex.roundtrip("XXXX-WXX-3");
        TestTimex.roundtrip("XXXX-12-05");
    }

    @Test
    public void dataTypesTimexRoundtripTime() {
        TestTimex.roundtrip("T17:30:45");
        TestTimex.roundtrip("T05:06:07");
        TestTimex.roundtrip("T17:30");
        TestTimex.roundtrip("T23");
    }

    @Test
    public void dataTypesTimexRoundtripDuration() {
        TestTimex.roundtrip("P50Y");
        TestTimex.roundtrip("P6M");
        TestTimex.roundtrip("P3W");
        TestTimex.roundtrip("P5D");
        TestTimex.roundtrip("PT16H");
        TestTimex.roundtrip("PT32M");
        TestTimex.roundtrip("PT20S");
    }

    @Test
    public void dataTypesTimexRoundTripNow() {
        TestTimex.roundtrip("PRESENT_REF");
    }

    @Test
    public void dataTypesTimexRoundtripDateTime() {
        TestTimex.roundtrip("XXXX-WXX-3T04");
        TestTimex.roundtrip("2017-09-27T11:41:30");
    }

    @Test
    public void dataTypesTimeRoundtripDateRange() {
        TestTimex.roundtrip("2017");
        TestTimex.roundtrip("SU");
        TestTimex.roundtrip("2017-WI");
        TestTimex.roundtrip("2017-09");
        TestTimex.roundtrip("2017-W37");
        TestTimex.roundtrip("2017-W37-WE");
        TestTimex.roundtrip("XXXX-05");
    }

    @Test
    public void dataTypesTimexRoundtripDateRangeStartEndDuration() {
        TestTimex.roundtrip("(XXXX-WXX-3,XXXX-WXX-6,P3D)");
        TestTimex.roundtrip("(XXXX-01-01,XXXX-08-05,P216D)");
        TestTimex.roundtrip("(2017-01-01,2017-08-05,P216D)");
        TestTimex.roundtrip("(2016-01-01,2016-08-05,P217D)");
    }

    @Test
    public void dataTypesTimexRoundtripTimeRange() {
        TestTimex.roundtrip("TEV");
    }

    @Test
    public void dataTypesTimexRoundtripTimeRangeStartEndDuration() {
        TestTimex.roundtrip("(T16,T19,PT3H)");
    }

    @Test
    public void dataTypesTimexRoundtripDateTimeRange() {
        TestTimex.roundtrip("2017-09-27TEV");
    }

    @Test
    public void dataTypesTimexRoundtripDateTimeRangeStartEndDuration() {
        TestTimex.roundtrip("(2017-09-08T21:19:29,2017-09-08T21:24:29,PT5M)");
        TestTimex.roundtrip("(XXXX-WXX-3T16,XXXX-WXX-6T15,PT71H)");
    }

    @Test
    public void dataTypesTimexToString() {
        Assert.assertEquals("5th May", new TimexProperty("XXXX-05-05").toString());
    }

    @Test
    public void dataTypesTimexToNaturalLanguage() {
        LocalDateTime today = LocalDateTime.of(2017, 10, 16, 0, 0);
        Assert.assertEquals("tomorrow", new TimexProperty("2017-10-17").toNaturalLanguage(today));
    }

    @Test
    public void dataTypesTimexFromTime() {
        Time time = new Time(23, 59, 30);
        Assert.assertEquals("T23:59:30", TimexProperty.fromTime(time).getTimexValue());
    }

    private static void roundtrip(String timex) {
        Assert.assertEquals(timex, new TimexProperty(timex).getTimexValue());
    }
}
